internal/storagenode/log_server.go

```diff
@@ -23,55 +24,161 @@ type logServer struct {
 
 var _ snpb.LogIOServer = (*logServer)(nil)
 
-func (ls logServer) Append(stream snpb.LogIO_AppendServer) (err error) {
-	req, rsp := &snpb.AppendRequest{}, &snpb.AppendResponse{}
+func (ls *logServer) Append(stream snpb.LogIO_AppendServer) error {
+	// Avoid race of Add and Wait of wgAppenders.
+	rt := ls.sn.mu.RLock()
+	ls.sn.wgAppenders.Add(2)
+	ls.sn.mu.RUnlock(rt)
+
+	cq := make(chan *logstream.AppendTask, ls.sn.appendPipelineSize)
+
+	go ls.appendStreamRecvLoop(stream, cq)
+
+	var eg errgroup.Group
+	eg.Go(func() error {
+		return ls.appendStreamSendLoop(stream, cq)
+	})
+	err := eg.Wait()
+	// The stream is finished by the client, which invokes CloseSend.
+	// That result from appendStreamSendLoop is nil means follows:
+	// - RecvMsg's return value is io.EOF.
+	// - Completion queue is closed.
+	// - AppendTasks in the completion queue are exhausted.
+	if err == nil {
+		ls.sn.wgAppenders.Done()
+		return nil
+	}
+
+	// Drain completion queue.
+	go ls.appendStreamDrainCQLoop(cq)
+
+	// The stream is finished by returning io.EOF after calling SendMsg.
+	if err == io.EOF {
+		return nil
+	}
+
+	var code codes.Code
+	switch err {
+	case verrors.ErrSealed:
+		code = codes.FailedPrecondition
+	case snerrors.ErrNotPrimary:
+		code = codes.Unavailable
+	default:
+		code = status.Code(err)
+		if code == codes.Unknown {
+			code = status.FromContextError(err).Code()
+		}
+
+	}
+	return status.Error(code, err.Error())
```

Comment:
How is this return code used?

Reply:
Not yet in the storage node client. We are still changing the code base to utilize the gRPC error codes.
For example, we currently use comparing strings to detect the sealed error, which will soon be fixed using gRPC error codes.
You can refer to #312.

Reply:
Does `verrors.ErrSealed` or `verrors.ErrNotPrimary` return to client as response. When LS is sealed, does `ls.appendStreamSendLoop(stream, cq)` fill response and return with `ErrSealed`?

Reply:
Yes, it does. The method `appendStreamSendLoop` returns an ErrSealed when the log stream is sealed. In that case, Append RPC replies to the client with an error coded `FailedPrecondition`, and the error string is "sealed". The client that receives a gRPC error coded `FailedPrecondition` and stringified with "sealed" parses the error string to detect ErrSealed. Do you have a follow-up question?

Reply:
When the LS is sealed, the gRPC stream appears to be closed. In my opinion, even if `ErrSealed` is delivered to the client through the response, the stream should be maintained. I hope this will be applied through other PRs.